### DISCUSSION TITLE
support pglite as datasource

### DISCUSSION THREAD

**Discussion by karlismelderis-mckinsey** [mikro-orm/mikro-orm#5945] [Ideas]
**Is your feature request related to a problem? Please describe.**
in tests [pglite](https://pglite.dev) feels to be great drop in replacement for postgres to avoid mocking or spinning up containers with real PG

**Describe the solution you'd like**
Have an option to use pglite as a datasource

**Additional context**
I'm new to MicroORM hence it's not clear where such implementation would start

**Comment by jaridmargolin** [comment]
@B4nan - Do you have any initial thoughts about supporting pglite? We are investigating different approaches to parallelizing our integration tests and pglite looked like an interesting solution!

**Reply by B4nan** [reply]
I gave it just a quick look, I guess we could modify the postgres knex dialect to use that if it has the same API. If not, we'd need to build a custom dialect. It's mostly about the knex support, the rest should be all reused, we don't work with the library directly outside of the type overrides.

**Reply by B4nan** [reply]
Looks like someone already made the knex dialect:

https://github.com/czeidler/knex-pglite

So this should be quite trivial to implement, I'll have a closer look once I start working on v6.4, still have some bugs to sort out first.

**Reply by ansarizafar** [reply]
Thats really good news. What is the ETA for v6.4?

**Reply by B4nan** [reply]
Hard to guess, no fixed timeline, but I actually got one fix locally that would be a better fit for v6.4, so it might be the next version already. In any case, it will take a few weeks.

**Reply by Jonovono** [reply]
Curious if you have looked any further in this @B4nan. Would love to give mikro orm a try in my Pglite app. In the meantime I am going to play around and see if I can get something working

**Reply by B4nan** [reply]
Not much, didn't have the energy to do this yet, but it should be really simple implementation-wise.

**Reply by B4nan** [reply]
I just learned that pg-lite is ESM, so native support for that will need to wait till next major.

**Comment by zachkirsch** [comment]
Can't wait for this!!

**Comment by ChrisJMacdonald** [comment]
Also looking forward to it!

**Comment by harryplusplus** [comment]
Hi there!
I'm a big fan of MikroORM.

I like PGlite, but since it’s not officially supported by MikroORM, I built this integration myself—on top of `@mikro-orm/postgresql`.

npm: [mikro-orm-pglite](https://www.npmjs.com/package/mikro-orm-pglite)
GitHub: [harryplusplus/mikro-orm-pglite](https://github.com/harryplusplus/mikro-orm-pglite)

I tried `knex-pglite` too, but MikroORM creates multiple Knex instances for database/schema checks, which doesn’t work well with PGlite’s in-memory model.
To solve this, PGliteConnection inside PGliteDriver now manages the PGlite instance directly, so you can use it easily—no extra setup.
If needed, you can inject your own PGlite instance to use things like file dbs, memory dump/load, cloning, or Postgres extensions.

Any feedback on the implementation is welcome!
Also, English isn’t my native language, so please let me know if anything is unclear—or if my wording could be improved. Thanks!

**Reply by mlhaufe** [reply]
How does this compare with:

npm: [@jadejr/mikro-orm-pglite](https://www.npmjs.com/package/@jadejr/mikro-orm-pglite)
GitHub: https://github.com/jadejr/mikro-orm-pglite

**Reply by harryplusplus** [reply]
Hi @mlhaufe,  
It seems that @jadejr/mikro-orm-pglite currently depends on @mikro-orm/core v7 (as seen in its [package.json](https://github.com/jadejr/mikro-orm-pglite/blob/6a94ad53577ec5e47d027db8b42726d5d747697f/package.json#L49) and the latest 1.0.3 release on npm).  
From what I can tell, it might be an early prototype exploring Kysely integration, since MikroORM v7 is moving from Knex to Kysely as the query builder (see discussion [#6116](https://github.com/mikro-orm/mikro-orm/discussions/6116)).

I’ve implemented a version that works with @mikro-orm/core 6.5.6 for current use cases.

**Reply by B4nan** [reply]
> moving from Knex to Kysely as the query builder

Query executor to be specific, query building is completely native in v7, no dependency needed.

I was wondering, is pglite a testing only solution (so it's fine if the package depends on `pg` too, since you'll use that in production anyway), or is it a standalone target and it would make more sense to have a driver without direct dependency on `pg`?

**Reply by harryplusplus** [reply]
(English is not my native language, and I received assistance from an AI translation tool.)

> Query executor to be specific, query building is completely native in v7, no dependency needed.

@B4nan Thank you for clarifying the MikroORM v7 feature distinction regarding query building. I appreciate the correction and am ready to address your questions about PGlite's status and driver dependencies.

> I was wondering, is pglite a testing only solution (...) or is it a standalone target ...

[PGlite](https://pglite.dev/) is designed as a standalone target, extending beyond a testing-only solution. This is evident through its support for persistence options: File system (Node/Bundle), IndexedDB (Browser), and In-memory storage.

> ... and it would make more sense to have a driver without direct dependency on `pg`?

I agree that a `pg`-independent driver is essential. Decoupling is necessary for PGlite's multi-platform nature to enable diverse use cases and full PostgreSQL feature support (like extensions) within MikroORM.

My implementation can be viewed as a Proof of Concept (POC) demonstrating feasibility. Architecturally, integration could effectively utilize `@mikro-orm/knex` commonization, similar to the approach taken for MySQL and SQLite. I welcome the opportunity to contribute to this integration effort.

**Reply by tdrz** [reply]
Hi @B4nan @harryplusplus 

I'm one of the maintainers of [PGlite](https://pglite.dev/). It would be great to have a MikroORM driver for it. 

Let me know if I can help!

**Reply by harryplusplus** [reply]
Hi @B4nan,

I’ve read the CONTRIBUTING.md guidelines and would like to share a proposal for an architecture enabling official PGlite support, following the recommended contribution process.

I’ve been exploring a clean approach for integrating PGlite without coupling to `pg` or `pg-query-stream`. Following the structure of `@mikro-orm/postgresql` and `@mikro-orm/knex`, here’s a possible implementation plan:
- `@mikro-orm/knex` additions
  - Add `PGliteKnexDialect.ts` extending `PostgreSqlKnexDialect` (overriding pg‑dependent logic).
  - Extract shared postgres logic into `BasePostgreSqlPlatform.ts`, `BasePostgreSqlSchemaHelper.ts`, and `BasePostgreSqlExceptionConverter.ts`.
- `@mikro-orm/postgresql` changes
  - Refactor current implementations to inherit from the new `BasePostgreSql*` classes (for backward compatibility).
- New `@mikro-orm/pglite` package
  - Add platform, schema helper, exception converter, driver, and connection classes extending the base PostgreSQL abstractions.
  - Add `PGliteMikroORM.ts` bootstrap file similar to `PostgreSqlMikroORM.ts`.
  
This structure allows PGlite to reuse most PostgreSQL logic while staying decoupled from native `pg` dependencies.

Would this direction make sense for a potential PR? I’d be happy to prepare a draft once you confirm it aligns with the project’s architecture and roadmap.

Thanks for your time and feedback! 🙏

**Reply by B4nan** [reply]
Yeah, makes sense, just like the sqlite drivers work. But let's target https://github.com/mikro-orm/mikro-orm/tree/v7, I wont be adding more drivers in the current stable, not to mention it feels unlikely this can work with a CJS project. v6 still supports node 18, where `require(esm)` fails.

**Reply by harryplusplus** [reply]
Hello @B4nan,
Thank you for your clear guidance. I’ll first try implementing and testing the PGlite integration against the `v7` branch, following a similar approach to the SQLite drivers.
Once it’s working properly, I’ll open a PR. I really appreciate your time and feedback.

**Comment by jadejr** [comment]
@harryplusplus : I think this should really wait for official pglite dialect in kysely before being merged. It's coming in .29. 

My mikro-orm driver was denied for upstream inclusion on the grounds of maintainership burden, so It would be surprising if one was accepted now.

My pglite driver was able to pass 99% of mikro-orm's postgres tests with some modifications . I still have most of changes I made to a v7 branch of mikro-orm.  If we could rework the connection parameter handling to be more flexible, it'd require even less modifications.    If you wanted to talk more about the specifics, we could talk on mikro-orm's discord.

Most of the modifications were related to result ordering and connection parameter handling. I did have to disable the replication and postgis tests.

**Reply by harryplusplus** [reply]
Thank you so much for your thorough reply, @jadejr.

@B4nan, I was wondering if it would still be appropriate for me to proceed with contributing to the official PGlite support on the v7 branch — in case I’ve missed any prior context or discussion along the way.

**Reply by B4nan** [reply]
If there will be an official dialect in kysely, I don't mind having an official driver here. I've changed my mind on this. I see how popular this request is, not many of the other feature requests are getting so much attention :] 

The problems with the client URL will need to be resolved either way, also because of other drivers. I believe I mentioned oracle earlier, they don't use regular URLs at all, and in the end, this will help with postgres too (e.g. #6892).

**Reply by harryplusplus** [reply]
Thank you so much for your openness and for reconsidering this, @B4nan.
I completely agree — every decision should naturally take into account the scale of the needs and the priority of the issues to be solved.

I also appreciate @jadejr’s initiative in taking the lead with a v7‑based PGlite implementation.

**Comment by jadejr** [comment]
@harryplusplus : Are you interested in helping with integrating pglite into the unit tests?

**Reply by harryplusplus** [reply]
@jadejr: Of course, I'd be happy to help with integrating PGlite into the unit tests. Where would you like me to start? I'm also a user of both MikroORM and PGlite, so the official integration is something I care deeply about, and I'd love to contrubite to the growth of the ecosystem.

**Reply by jadejr** [reply]
Currently all the connection parameters for the pg unit tests are hardcoded so they'd need to be made more dynamic.

How dynamic they end up being depends on what part of the code we really want to exercise. 

We have 3 different things we can think about just for deciding how we connect and work with pglite

1. Using the pglite driver itself with on-disk storage
2. Using the pglite driver completely in memory 
3. Connecting with a pglite socket server (this didn't exist when i first started my integration). This one just uses the existing postgres driver.

The first require things to be much more dynamic when it comes to the connection parameters The second changes test assumptions.  All the tests assume the the postges server exists on disk unlike with sqlite where we see both approaches covered, with the in-memory one being most used.  The simplest approach is to just continue assuming on disk storage the whole time, but with just a few added tests for testing the memory related connection parameters.

As far as the connection parameters themselves go though, that's where we have a problem. B4nan mentioned the main one related to client urls. Currently there is an assumption that clientUrls need a `pathname` and the  `pathname`  is the same as `dbName`. This is incorrect in both usages of pglite. Our pathame (if we have one) needs to be the `dataDir`. It could also be something like `dataDir/dbName`, but it wasn't worth doing due to the other constraints atm.
An unrelated side benefit of fixing this could mean that sqlite with ATTACH DB could be made to work with mikro-orm.

My current hacky approach uses `pglite://?dbName=my_db` for in memory connections with no user on the simpler side, and with on disk storage I go with `pglite://ANYUSER@localhost/path/to/dataDir?dbName=my_db` but I'm open for changes here. 

Another thing related to connection parameters, is that my driver currently uncoditionally loads `citext` and `tsvector` to support the tests, but they should really be made optional and would need to loaded via `driverOptions` for the relevant tests. 

As far as actual test content changes go, here are the ones I can remember off the top of my head.

1. The driver itself will need need to be loaded conditonally (either pglite or postgresql)
2. postgis tests are skippable (until pglite restores this functionality)
3. Any tests with `REPLICA` fails for reasons I don't yet understand. There aren't that many though. They could be skippable for pglite
4. The ordering of various query results are not the same between different drivers. I looked into this, and result ordering is never guaranteed by spec. This means some of the test queries need explicit `ORDER BY`/`orderBy`
5. PGlite is much more limited when it comes to locale/collation support to keep the size reasonable. I don't think this affects any of the current tests, but it could affect future ones.


There's one more major issue which I'm having trouble properly elucidating right now and it is related to handling multiple statements. This is a bit of a brain dump from here on.  Pglite DOES support handling multiple statements but the driver interface both on the kysely and pglite sides make it more annoying than it should be.  pglite offers `exec` as a separate method rather than the much nicer way pg does.  I hacked around this by adding this to the schema generator `execute` method

```ts
// ... snipped. You can see the whole implementation in PgLiteSchemaGenerator.ts
for (const group of groups) {
      const query = group.join('\n');
      // @todo: add to type if we keep the loadQuery concept
      if ('loadQuery' in this.driver && typeof (this.driver as unknown as PgLiteDriver).loadQuery === 'function') {
        await (this.driver as unknown as PgLiteDriver).loadQuery(query);
      } else {
        throw new Error('PgLiteDriver with loadQuery method is required');
      }
    }
```
which ends up calling pglite's exec()` 

I then adjusted some of the tests to call `schema.execute()` instead of `connection.execute()`. 

We could make all the tests use single line statements to avoid this (or enhance the statement splitting) and just turn `supportsMultipleStatements` off, but I'd prefer if we could do better.

**Comment by jadejr** [comment]
Some of this connector paramter stuff I'm speaking of could be just as well applied to supporting the various sqlite drivers as well. Currently mikro-orm does not support node's built in sqlite driver for example. I still have some sqlite migration stuff in my codebase, so I ended up going with `https://github.com/subframe7536/kysely-sqlite-tools and created my own mini node:sqlite driver for migration purposes since kysely itself has not upstreamed node:sqlite support yet.

**Reply by B4nan** [reply]
sqlite in node is still experimental, last time i checked, it prints a warning when used, that's not something i would like to hurry with supporting (and it is trivial to support it yourself, everything you need is exposed already because of libsql driver).

**Comment by harryplusplus** [comment]
> I think this should really wait for official pglite dialect in kysely before being merged. It's coming in .29.

@jadejr: I agree that it makes sense to proceed with the actual implementation (draft PR) at the timing you mentioned. I have confirmed that the official Kysely PGlite dialect accepts a PGlite instance as part of its structure, so I have no additional concerns.

For reference, see the Kysely next branch: https://github.com/kysely-org/kysely/pull/1583
And specifically: [pglite-dialect-config.ts#L22](https://github.com/kysely-org/kysely/blob/4eb6e8b5799e3a6c5610a7a81c5c6e2da3b86ee0/src/dialect/pglite/pglite-dialect-config.ts#L22)

Thank you very much for providing such a detailed explanation of the implementation.

> 1. Using the pglite driver itself with on-disk storage
> 2. Using the pglite driver completely in memory

From the library provider’s point of view, I believe that supporting the following two use cases—an in-memory store and a disk-based store—and writing tests for them is the correct approach.

> 3. Connecting with a pglite socket server (this didn't exist when i first started my integration). This one just uses the existing postgres driver.

Regarding the use cases you mentioned, I believe that the case involving [PGlite Socket](https://pglite.dev/docs/pglite-socket﻿) does not need to be supported by the official MikroORM PGlite driver (tentatively named @mikro-orm/pglite). This is likely because PGlite Socket exposes the PostgreSQL protocol interface rather than the PGlite instance interface. From MikroORM’s perspective, @mikro-orm/postgresql already uses pg (node-postgres) and can handle this use case—as shown in the example [here](https://pglite.dev/docs/pglite-socket#using-node-js-clients﻿).

> The simplest approach is to just continue assuming on disk storage the whole time, but with just a few added tests for testing the memory related connection parameters.

I agree. Although nothing has been implemented yet, based on my review of the code, it seems best to create test suites for the common PostgreSQL drivers (@mikro-orm/postgresql, @mikro-orm/pglite) as well as specialized test suites for specific cases like MySQL and SQLite. While I believe specialized test suites should be minimized, following the typical 80/20 rule, it is unlikely they can be entirely avoided.

> As far as the connection parameters themselves go though, that's where we have a problem.

I also deeply relate to **the client URL issue** mentioned above, and here are my thoughts:

The clientUrl should be optional and usable even when undefined. In the case of PGlite, it is expected to operate with an in-memory storage by default.

The direction of this implementation still needs further confirmation. Simplistically, removing validation at the ORM level would require delegating validation to the lower driver levels, meaning each driver must implement it. I need to review the code more to determine if validation at the ORM level is truly necessary.

Regarding connection parameters, here are my views (still open for discussion):

edit) As a precondition, if the overrides (driverOptions) passed to the createKyselyDialect(overrides) method does not have the pglite property...

- In MikroORM, `clientUrl` corresponds to PGlite’s `dataDir`.
- `driverOptions` in MikroORM maps to `options` in PGlite.
- If `clientUrl` is defined in MikroORM, it takes precedence over `driverOptions.dataDir` because in PGlite, `dataDir` in options is lower priority than the first argument.

This is based on my understanding after reviewing the [PGlite constructor interface](https://pglite.dev/docs/api#main-constructor), which allows two initialization methods: `dataDir: string, options: PGliteOptions` or just `options: PGliteOptions`. The `dataDir` property in `options` is optional and specifies the directory for storing the Postgres database when not passed as the first argument.

From a user perspective, it would be preferable not to introduce a separate MikroORM-specific clientUrl rule to avoid extra learning and maintain consistency. Passing it through as-is, and adding non-mutating precautions only if common pitfalls occur, seems best. Examples of such pitfalls include throwing exceptions for unsupported browser storage schemes like idb:// or ensuring directory existence for filesystem paths. However, detailed validation discussion should follow after robust official support is implemented.

> Another thing related to connection parameters, is that my driver currently uncoditionally loads citext and tsvector to support the tests, but they should really be made optional and would need to loaded via driverOptions for the relevant tests.

I understand this well. Thank you for the detailed explanation.

> As far as actual test content changes go, here are the ones I can remember off the top of my head. ...

I will verify implementation details during actual testing. Thank you very much for the thorough explanation despite the significant time since the implementation.

--- 

> The problems with the client URL will need to be resolved either way, also because of other drivers. I believe I mentioned oracle earlier, they don't use regular URLs at all, and in the end, this will help with postgres too (e.g. https://github.com/mikro-orm/mikro-orm/discussions/6892).

I will verify that MikroORM v7 supports this after the Kysely 0.29 release. For now, as @B4nan shared, I plan to review how clientUrl is implemented in v7 and submit a draft PR if needed.

**Reply by B4nan** [reply]
> If clientUrl is defined in MikroORM, it takes precedence over driverOptions.dataDir

Nope, anything in `driverOptions` has precedence, that's the whole idea behind `driverOptions`, to allow overriding what the ORM maps automatically. If the pglite option is called `dataDir`, the ORM would map `clientUrl` to it, and if you would provide `driverOptions.dataDir`, that would override it.

Btw you could already use kysely 0.29, they provide dev builds for it, it would be enough for the PR itself, and we could only wait for the release with the actual merge. There might be some non-trivial git conflicts, since I will be rebasing v7 branch once in a while (currently waiting for #6939, will rebase once we merge it). I usually just squash all the changes, stash them and hard reset to the latest upstream version instead of battling git rebase, that way it's pretty simple.

**Reply by harryplusplus** [reply]
That is a flexible system. Good software design. And I really like that method.

**Comment by jadejr** [comment]
> . From MikroORM’s perspective, @mikro-orm/postgresql already uses pg (node-postgres) and can handle this use case—as shown in the example [here](https://pglite.dev/docs/pglite-socket#using-node-js-clients%EF%BB%BF).

This is what i thought i said. but it still needs it's own connection string  since it doesn't support all same clientUrl options as pg does.

I know you said you used AI to try to translate for you, but this entirely too much text. please make it more brief.


I've already ran the test suite so i know it can pass with those minimal modifcations I mentioned. I just don't have them in public anywhere yet since they got broken oncei split out my driver and are probably relatively trivial to reimplement anyways.

**Reply by harryplusplus** [reply]
I apologize for the lengthy message. Thank you for your understanding.

edit) I tried not to sound rude and ended up writing too much. 😅

**Reply by harryplusplus** [reply]
Are you on Discord? I'd love to discuss the detailed implementation plans there.

**Comment by julianpung** [comment]
Is this still being worked on? Would love to see a pglite integration.

**Reply by B4nan** [reply]
There is some third-party driver for pglite (maybe even more than one). Native support will come to v7 once kysely 0.29 it out (it will add a pglite dialect).

**Reply by jadejr** [reply]
I have 

> Is this still being worked on? Would love to see a pglite integration.

I've been waiting for 0.29 to b e released before bothering to try to upstream mine, since the dialect i based it on does entirely too much.

**Comment by B4nan** [comment]
#7622 will land in v7.1, let's hope kysely 0.29 will ship stable version soon
